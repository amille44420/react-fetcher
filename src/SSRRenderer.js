export default class SSRRenderer {
  constructor(initialState = []) {
    // list of promises to wait for
    this.promises = [];

    // persisted data through multiple run for SSR
    this.state = initialState;

    // current index
    this.index = 0;

    // has the state been consumed
    this.consumed = false;
  }

  hasPromises() {
    return this.promises.length > 0;
  }

  hasState(index) {
    return !this.consumed && index + 1 <= this.state.length;
  }

  register() {
    const { index } = this;
    this.index += 1;

    return index;
  }

  getState(index) {
    const state = this.state[index];

    if (index + 1 === this.state.length) {
      this.consumed = true;
    }

    return state;
  }

  attachPromise(promise, props, index) {
    this.promises.push(
      promise.then(result => {
        // register everything for this promise
        this.state[index] = { result, props };

        // and return the result again
        return result;
      }),
    );
  }

  consume() {
    // get promises to wait for
    const { promises } = this;

    // reset the renderer
    this.index = 0;
    this.promises = [];
    this.consumed = false;

    return Promise.all(promises);
  }
}
