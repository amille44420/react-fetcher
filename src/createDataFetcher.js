import React, {
  PureComponent,
  Component,
  createContext,
  memo,
  useContext,
} from 'react';
import { isValidElementType } from 'react-is';
import shallowequal from 'shallowequal';
import SSRContext from './SSRContext';

const fetchDefaultOptions = {
  // fetcher name
  name: 'data',
  // we may provide options for the providers
  providerOptions: null,
  // we may provide options for the consumers
  consumerOptions: null,
};

const providerDefaultOptions = {
  // variables (function or object)
  variables: null,
  // pure option
  pure: true,
  // should we refetch on props change
  refetchOnPropChanges: false,
  // function defining if there's a change on props
  shouldUpdate: shallowequal,
  // we may want to process the fetched data
  cleanData: null,
};

const consumerDefaultOptions = {
  // pure option
  pure: true,
};

const getDisplayName = WrappedComponent =>
  WrappedComponent.displayName || WrappedComponent.name || 'Component';

const createDataFetcher = (fetcher, customFetcherOptions = null) => {
  // create final options
  const options = { ...fetchDefaultOptions, ...customFetcherOptions };

  // capitalize the name to use it as display name
  const displayName =
    options.name.charAt(0).toUpperCase() + options.name.substr(1);

  // create a context for this data fetcher
  const Context = createContext(null);

  // we build a custom hook to consume the context
  const useData = () => {
    return useContext(Context);
  };

  // we build the HOC to fetch data and provide the context
  const withDataFetcher = providerCusomOptions => {
    // initialize the options with the default options
    let providerOptions = {
      ...providerDefaultOptions,
      ...options.providerOptions,
    };

    // here is our factory to build the HOC provider
    const getHOC = WrappedComponent => {
      const ComponentBase = options.pure ? PureComponent : Component;

      class Provider extends ComponentBase {
        static contextType = SSRContext;

        constructor(props) {
          super(props);

          // pending promise to fetch data
          this.fetchPromise = null;
          // previous props & state
          this.prevProps = null;
          // current context
          this.contextValue = null;
          // SSR index
          this.index = null;
        }

        getVariables() {
          return providerOptions.variables instanceof Function
            ? providerOptions.variables(this.props)
            : providerOptions.variables;
        }

        async fetch(variables) {
          const result = { err: null, data: null };

          try {
            // inform we're fetching
            this.updateContext({ loading: true });

            // update the state at first
            const data = await fetcher(variables, this.props);

            const cleanedData = providerOptions.cleanData
              ? providerOptions.cleanData(data, variables, this.props)
              : data;

            // update the state with the data
            this.updateContext({
              loading: false,
              data: cleanedData,
              error: null,
            });
            // do an update just after fetching
            this.forceUpdate();

            // update the result
            result.data = cleanedData;
          } catch (error) {
            // update the state with the error
            this.updateContext({ loading: false, data: null, error });
            // do an update after getting an error
            this.forceUpdate();

            // update the result
            result.error = error;
          } finally {
            this.fetchPromise = null;
          }

          return result;
        }

        refetch = (customVariables = null) => {
          const variables = {
            // first get variables from props
            ...this.getVariables(),
            ...customVariables,
          };

          // fetch again
          this.fetchPromise = this.fetch(variables, true);
          // and force the update
          this.forceUpdate();

          return this.fetchPromise;
        };

        get shouldFetch() {
          if (this.contextValue === null && this.fetchPromise === null) {
            // first fetch
            return true;
          }

          if (this.prevProps === this.props || this.fetchPromise) {
            // props are the same or it's already fetching
            return false;
          }

          // only fetch if we're looking to fetch on prop updates
          return (
            providerOptions.refetchOnPropChanges &&
            !providerOptions.shouldUpdate(this.props, this.prevProps)
          );
        }

        updateContext(newContext) {
          // the context
          this.contextValue = {
            // initial values
            error: null,
            data: null,
            // override with the current context
            ...this.contextValue,
            // override again with the new context above it
            ...newContext,
            // and ensure the right refetch method is given
            refetch: this.refetch,
          };
        }

        render() {
          const { ssr, renderer } = this.context;

          if (renderer) {
            this.index = renderer.register();

            if (renderer.hasState(this.index)) {
              const { result, props } = renderer.getState(this.index);
              this.updateContext({ loading: false, ...result });
              this.prevProps = props;
            }
          }

          if (this.shouldFetch) {
            // compute variables from props
            const variables = { ...this.getVariables() };
            // and fetch
            this.fetchPromise = this.fetch(variables);
            // remember the props we used
            this.prevProps = this.props;

            if (ssr) {
              // register our fetch to be processed on SSR
              renderer.attachPromise(this.fetchPromise, this.props, this.index);
            }
          }

          return (
            <Context.Provider value={this.contextValue}>
              <WrappedComponent {...this.contextValue} {...this.props} />
            </Context.Provider>
          );
        }
      }

      const wrappedDisplayName = getDisplayName(WrappedComponent);

      Provider.displayName = `with${displayName}Fetcher(${wrappedDisplayName})`;

      return Provider;
    };

    if (isValidElementType(providerCusomOptions)) {
      // the argument is a valid react element type, so let's consider there's no options
      return getHOC(providerCusomOptions);
    }

    // we do have custom options, let's merge those
    providerOptions = { ...providerOptions, ...providerCusomOptions };

    // and return the factory
    return getHOC;
  };

  // then the HOC to consume data
  const withData = consumerCustomOptions => {
    // initialize the options with the default options
    let customerOptions = {
      ...consumerDefaultOptions,
      ...options.consumerOptions,
    };

    // here is our factory to build the HOC consumer
    const getHOC = WrappedComponent => {
      let Consumer = props => {
        const contextValue = useData();

        return <WrappedComponent {...contextValue} {...props} />;
      };

      if (customerOptions.pure) {
        Consumer = memo(Consumer);
      }

      const wrappedDisplayName = getDisplayName(WrappedComponent);

      Consumer.displayName = `with${displayName}(${wrappedDisplayName})`;

      return Consumer;
    };

    if (isValidElementType(consumerCustomOptions)) {
      // the argument is a valid react element type, so let's consider there's no options
      return getHOC(consumerCustomOptions);
    }

    // we do have custom options, let's merge those
    customerOptions = { ...customerOptions, ...consumerCustomOptions };

    // and return the factory
    return getHOC;
  };

  return {
    [`use${displayName}`]: useData,
    [`with${displayName}Fetcher`]: withDataFetcher,
    [`with${displayName}`]: withData,
  };
};

export default createDataFetcher;
