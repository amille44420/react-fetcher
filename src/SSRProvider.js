import React, { useMemo, useContext } from 'react';
import PropTypes from 'prop-types';
import SSRContext from './SSRContext';
import SSRRenderer from './SSRRenderer';

const SSRProvider = ({ state, children }) => {
  let context = useContext(SSRContext);

  context = useMemo(
    () => ({
      ...context,
      renderer: new SSRRenderer(state),
    }),
    [state, context],
  );

  return <SSRContext.Provider value={context}>{children}</SSRContext.Provider>;
};

SSRProvider.propTypes = {
  state: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  children: PropTypes.node.isRequired,
};

export default SSRProvider;
