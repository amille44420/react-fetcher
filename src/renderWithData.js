import { createElement } from 'react';
import SSRContext from './SSRContext';
import SSRRenderer from './SSRRenderer';

const renderWithData = (tree, render) => {
    const renderer = new SSRRenderer();

    const process = () => {
        // wrap the tree with our provider to dispatch our SSR context
        const element = createElement(SSRContext.Provider, { value: { ssr: true, renderer } }, tree);

        // call the rendering method
        const output = render(element);

        // loop until there's no more promises
        return renderer.hasPromises() ? renderer.consume().then(process) : [output, renderer.state];
    };

    // start the loop
    return process();
};

export default renderWithData;
