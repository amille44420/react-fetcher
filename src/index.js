import createDataFetcher from './createDataFetcher';
import renderWithData from './renderWithData';
import SSRProvider from './SSRProvider';

export { createDataFetcher, renderWithData, SSRProvider };
