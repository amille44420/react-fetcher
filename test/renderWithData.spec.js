import React from 'react';
import { renderToString } from 'react-dom/server';
import renderer from 'react-test-renderer';
import { createDataFetcher, renderWithData, SSRProvider } from '../src/index';

test('test renderWithData', async () => {
  // prepare a variable in which we'll store the resolver
  let resolvePromise = null;

  // fetcher function persisting the resolve method
  const fetcher = jest.fn(
    () =>
      new Promise(resolve => {
        resolvePromise = resolve;
      }),
  );

  // create our fetcher to test
  const { withDataFetcher, useData } = createDataFetcher(fetcher);

  // inner component to run for our tests
  const Inner = () => {
    const { loading, data } = useData();

    return !loading && <p>{data}</p>;
  };

  // apply the HOC
  const App = withDataFetcher(Inner);

  // render our tree
  const renderPromise = renderWithData(<App />, renderToString);

  // we expect html to be a promise
  expect(renderPromise).toBeInstanceOf(Promise);
  // we also expected resolvePromise to be a function (resolve method)
  expect(resolvePromise).toBeInstanceOf(Function);

  // resolve our promise
  resolvePromise(42);
  const [html, state] = await renderPromise;
  expect(fetcher.mock.calls.length).toBe(1);
  expect(html).toMatchSnapshot();
  expect(state).toMatchSnapshot();

  // try to render it again while injecting the state
  const injectedRenderer = renderer.create(
    <SSRProvider state={state}>
      <App />
    </SSRProvider>,
  );

  // we shouldn't have any call on our fetcher
  expect(fetcher.mock.calls.length).toBe(1);
  expect(injectedRenderer.toJSON()).toMatchSnapshot();
});

test('test renderWithData with nested fetchers', async () => {
  // create our first fetcher
  const firstFetcher = jest.fn(() => 42);
  const { withFirstFetcher, useFirst } = createDataFetcher(firstFetcher, {
    name: 'first',
  });
  // create our second fetcher
  const secondFetcher = jest.fn((variables, { value = 0 }) => value + 1);
  const { withSecondFetcher, useSecond } = createDataFetcher(secondFetcher, {
    name: 'second',
  });

  // create components
  let SecondLayout = ({ value }) => {
    const { loading, data } = useSecond();

    return (
      !loading && (
        <p>
          first {value} then {data}
        </p>
      )
    );
  };

  SecondLayout = withSecondFetcher({ refetchOnPropChanges: true })(
    SecondLayout,
  );

  let FirstLayout = () => {
    const { loading, data } = useFirst();

    return <SecondLayout value={loading ? data : 0} />;
  };

  FirstLayout = withFirstFetcher(FirstLayout);

  // render it
  const [html, state] = await renderWithData(<FirstLayout />, renderToString);

  // ensure we didn't call the fetchers less or more than expected
  expect(firstFetcher.mock.calls.length).toBe(1);
  expect(secondFetcher.mock.calls.length).toBe(2);

  // check the output
  expect(html).toMatchSnapshot();
  expect(state).toMatchSnapshot();

  // try to render it again while injecting the state
  const injectedRenderer = renderer.create(
    <SSRProvider state={state}>
      <FirstLayout />
    </SSRProvider>,
  );
  // we shouldn't have any call on our fetcher
  expect(firstFetcher.mock.calls.length).toBe(1);
  expect(secondFetcher.mock.calls.length).toBe(2);
  expect(injectedRenderer.toJSON()).toMatchSnapshot();
});
