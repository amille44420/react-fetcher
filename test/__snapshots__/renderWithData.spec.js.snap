// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test renderWithData 1`] = `"<p>42</p>"`;

exports[`test renderWithData 2`] = `
Array [
  Object {
    "props": Object {},
    "result": Object {
      "data": 42,
      "err": null,
    },
  },
]
`;

exports[`test renderWithData 3`] = `
<p>
  42
</p>
`;

exports[`test renderWithData with nested fetchers 1`] = `"<p>first <!-- -->0<!-- --> then <!-- -->1</p>"`;

exports[`test renderWithData with nested fetchers 2`] = `
Array [
  Object {
    "props": Object {},
    "result": Object {
      "data": 42,
      "err": null,
    },
  },
  Object {
    "props": Object {
      "value": 0,
    },
    "result": Object {
      "data": 1,
      "err": null,
    },
  },
]
`;

exports[`test renderWithData with nested fetchers 3`] = `
<p>
  first 
  0
   then 
  1
</p>
`;
